flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,387 @@
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus) {
+					return jobStatus == JobStatus.FINISHED;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.FINISHED, jobStatusFuture.get());
+
+		// make sure we saw a successful restore
+		successfulRestoreLatch.await();
+
+		assertThat("We saw illegal restores.", CheckpointBlockingFunction.illegalRestores.get(), is(0));
+	}
+
+	/**
+	 * Requests the {@link JobStatus} of the job with the given {@link JobID}.
+	 */
+	private Future<JobStatus> getJobStatus(
+		final ActorGateway jobManager,
+		final JobID jobId,
+		final FiniteDuration timeout) {
+
+		scala.concurrent.Future<Object> response =
+			jobManager.ask(JobManagerMessages.getRequestJobStatus(jobId), timeout);
+
+		FlinkFuture<Object> flinkFuture = new FlinkFuture<>(response);
+
+		return flinkFuture.thenApply(new ApplyFunction<Object, JobStatus>() {
+			@Override
+			public JobStatus apply(Object value) {
+				if (value instanceof JobManagerMessages.CurrentJobStatus) {
+					return ((JobManagerMessages.CurrentJobStatus) value).status();
+				} else if (value instanceof JobManagerMessages.JobNotFound) {
+					throw new RuntimeException(
+						new IllegalStateException("Could not find job with JobId " + jobId));
+				} else {
+					throw new RuntimeException(
+						new IllegalStateException("Unknown JobManager response of type " + value.getClass()));
+				}
+			}
+		});
+	}
+
+	private static class UnboundedSource implements SourceFunction<String> {
+		private boolean running = true;
+
+		@Override
+		public void run(SourceContext<String> ctx) throws Exception {
+			while (running) {
+				ctx.collect("hello");
+				// don't overdo it ... ;-)
+				Thread.sleep(50);
+				if (CheckpointBlockingFunction.afterMessWithZooKeeper.get()) {
+					break;
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			running = false;
+		}
+	}
+
+	private static class CheckpointBlockingFunction
+			extends RichMapFunction<String, String>
+			implements CheckpointedFunction {
+
+		// verify that we only call initializeState()
+		// once with isRestored() == false. All other invocations must have isRestored() == true. This
+		// verifies that we don't restart a job from scratch in case the CompletedCheckpoints can't
+		// be read.
+		static AtomicInteger allowedInitializeCallsWithoutRestore = new AtomicInteger(1);
+
+		// we count when we see restores that are not allowed. We only
+		// allow restores once we messed with the HA directory and moved it back again
+		static AtomicInteger illegalRestores = new AtomicInteger(0);
+		static AtomicInteger successfulRestores = new AtomicInteger(0);
+
+		// whether we are after the phase where we messed with the ZooKeeper HA directory, i.e.
+		// whether it's now ok for a restore to happen
+		static AtomicBoolean afterMessWithZooKeeper = new AtomicBoolean(false);
```

Comment:
I was initially confused why all of these are Atomic booleans/integers, etc. Could they not be simple volatile variables? Not that the performance matters, but I was just wondering whether something here assumes atomicity.